javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/matcher/ClassLoaderHasClassesNamedMatcher.java

```diff
@@ -46,4 +59,41 @@ private boolean hasResources(ClassLoader cl) {
     }
     return true;
   }
+
```

Comment:
I think this is a step in the right direction.

**My interpretation:**  (Please confirm if I'm interpreting this correctly)
So the new Manger is holding a list of all matchers, which correspond to N instrumentations.   For each matcher  a BitSet of size M will eventually  expand out.  That is N bitsets of length M being crated.    Previously, N instrumentations were each creating M cache entries, one entry for every ClassLoader instance encountered.   Each entry in the WeakCache creates a weak key, so previously the total weak keys created amounted to N x M.  Now, the a single weak key is created for each instrumenation.  This is a very good improvement.  During working on PR #7710, new instances of the Janino classloader were created in a runaway fashion inside my legacy application, counting for millions of classloader-to-boolean weakkeys.   Here, we will end up with just a bunch of  very long BitSets if a classloader storm happens.   

_Previous_
Trace 5Mil classloaders took 5Mil Boolean instances ~ 40MB x 112instumentations ~ 4.5GB (but my heap dumps only consumed 150MB since JDK 8 reuses Boolean instances)

_Now_ 
Track 5Mil ClassLoaders with BitSet will be 78KB per instrumentation ~ 78KB x 112 ~ 8Mb   **MUCH BETTER**

(Based my BitSet calculation on internals of a bit set modeled with array of Long[].  5Mil/64 = 78k Longs)

Reply:
In my opinion you analysis is flawed. It is based on the need to keep track of 5M class loaders, but that situation only arose when cache keys were not removed due to a bug. Given such bug the code in this pr will eventually also fail, it will just take a bit longer.

>  Here, we will end up with just a fixed list of very long BitSets if a classloader storm happens.

The size of the bit set is based on the count of the matchers not class loaders. This size is deterministic, I think it was a bit over 200. There is a bit set for each class loader.

> Trace 5Mil classloaders took 5Mil Boolean instances ~ 40MB x 112instumentations ~ 4.5GB (but my heap dumps only consumed 150MB since JDK 8 reuses Boolean instances)

A new `Boolean` instance is created only when you call `new Boolean(true)`. Boxing conversion calls `Boolean valueOf(boolean b)` (didn't verify this) which does not create new instances. I wouldn't expect these Boolean values to cause problems. My guess would be that a large consumer would be the array that backs the map. If you have 5M keys then the backing map would have 8M elements. And even large would be the map keys. Minimum size of a java object is like 16 bytes, with 5M keys that alone would be 80M. In my opinion it is better to build a small demo app that replicates the situation and let Eclipse memory analyzer compute the retained size.